File: app/src/main/kotlin/com/king/zxing/app/PaymentQrClassifier.kt
package com.king.zxing.app

import java.net.URI
import java.util.Locale

/** Strict payment-QR classification; normal web URLs must remain normal results. */
object PaymentQrClassifier {
    enum class Provider { WECHAT, ALIPAY, PAYPAL }

    fun classify(raw: String?): Provider? {
        val value = raw?.trim().orEmpty()
        if (value.isEmpty()) return null
        val lower = value.lowercase(Locale.ROOT)
        if (lower.startsWith("wxp://") || lower.startsWith("weixin://wxpay/")) {
            return Provider.WECHAT
        }
        val uri = runCatching { URI(value) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase(Locale.ROOT) ?: return null
        val host = uri.host?.lowercase(Locale.ROOT) ?: return null
        val path = uri.path.orEmpty()
        if (scheme in setOf("http", "https") && host == "qr.alipay.com" && path.length > 1) {
            return Provider.ALIPAY
        }
        if (scheme in setOf("http", "https") && isPayPalHost(host) && isPayPalPaymentPath(host, path)) {
            return Provider.PAYPAL
        }
        return null
    }

    private fun isPayPalHost(host: String): Boolean =
        host == "paypal.me" || host == "www.paypal.me" ||
            host == "paypal.com" || host == "www.paypal.com"

    private fun isPayPalPaymentPath(host: String, path: String): Boolean = when (host) {
        "paypal.me", "www.paypal.me" -> path.length > 1
        else -> path.startsWith("/qrcodes/p2pqrc/", ignoreCase = true)
    }
}

File: swiftui/ZXingX/Models/PaymentQrClassifier.swift
//
//  PaymentQrClassifier.swift
//  ZXingX (iOS)
//
//  Port of Android PaymentQrClassifier.kt — strict payment-QR classification.
//  Normal web URLs must remain normal results.
//

import Foundation

enum PaymentQrClassifier {

    enum Provider {
        case wechat
        case alipay
        case paypal
    }

    static func classify(_ raw: String?) -> Provider? {
        let value = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if value.isEmpty { return nil }
        let lower = value.lowercased()
        if lower.hasPrefix("wxp://") || lower.hasPrefix("weixin://wxpay/") {
            return .wechat
        }
        // java.net.URI(value) throws on malformed input → null; URLComponents
        // is the closest Foundation equivalent (scheme/host/path extraction).
        guard let components = URLComponents(string: value),
              let scheme = components.scheme?.lowercased(),
              let host = components.host?.lowercased() else {
            return nil
        }
        let path = components.path
        if (scheme == "http" || scheme == "https"),
           host == "qr.alipay.com", path.count > 1 {
            return .alipay
        }
        if (scheme == "http" || scheme == "https"),
           isPayPalHost(host), isPayPalPaymentPath(host: host, path: path) {
            return .paypal
        }
        return nil
    }

    private static func isPayPalHost(_ host: String) -> Bool {
        host == "paypal.me" || host == "www.paypal.me" ||
            host == "paypal.com" || host == "www.paypal.com"
    }

    private static func isPayPalPaymentPath(host: String, path: String) -> Bool {
        switch host {
        case "paypal.me", "www.paypal.me":
            return path.count > 1
        default:
            return path.lowercased().hasPrefix("/qrcodes/p2pqrc/")
        }
    }
}
